### ISSUE TITLE
Repo dead?

### ISSUE THREAD

**Issue by rubytastic** [arut/nginx-rtmp-module#605]
Are there plans to pickup this repo after months of inactivity?

**Comment by jordanbaucke** [comment]
I don't know arut bumped the nginx compatible build # a few days ago: https://github.com/arut/nginx-rtmp-module/commit/f62a0838064baf089ad9fe3c8f6f2ffe0775afb2

**Comment by mknwebsolutions** [comment]
Arut is working directly for nginx now.

**Comment by misiek08** [comment]
Yeah, money will destroy every good OS project. Started with HDS, slowly dying now.

Why I'm not helping? I just don't fully understand nginx model, so I'm forking alternative rtmp server and working on it.

**Comment by mknwebsolutions** [comment]
I think the repo as it stand is still quite powerful, you just really need to understand it inside and out.  I've been working on it daily since November 2014 and have a full scale application out of it.  I do know that Arut is still committing time to time to it though.

**Comment by jordanbaucke** [comment]
@mknwebsolutions would like to hear more about the size and scope of your application (if just generalities) some interesting aspects I'd like to learn more about at scale: load balancing, memory and bandwidth consumption, encoding / re-encoding at scale. Thanks!

**Comment by mknwebsolutions** [comment]
@jordanbaucke the quick summary is: rtmp output scaling absolutely sucks, thousands of streams later and we realized that just no one runes rtmp outputs.  Scaling and relaying from the top level seemed simple but there's just so many moving parts that can break rtmp.  Use RTMP for input, but use HLS for output.  Twitch, Youtube, Xfinity / Cable companies, etc all use HLS. 

With that being said, nginx rtmp should be easy to use since it's now on your broadcasters to stream a clean connection to it, HLS (standard http req) does the rest.  We use Amazon's CDNs to push HLS out for us.  Since we've made the move my stress level has dropped by at least 60%.

**Comment by jordanbaucke** [comment]
@mknwebsolutions thanks! That was my assumption - ingest via RTMP, re-encode and then propagate out HLS versions via a CDN *(the tool that's creating the broadcast, say a Flash component generates RTMP - are their tools that broadcast directly in HLS?)*

As an aside, when you distribute your HLS content to these CDNs there isn't really any magic right? Anywhere I could read about this - I understand writing the HLS components to a local disc, do you just point the CDN to a relative URL and it reads and takes care of the rest? Or do you push the HLS playlist / files out to them?

**Comment by mknwebsolutions** [comment]
@jordanbaucke bingo, exactly.  There's definitely a few other out there i.e. wowza, nodejs, etc... I def prefer nginx rtmp though.

Correct on CDN too!  Few ways to do it, my route entails pushing the files to amazon S3 via S3FS.  The playlist and files are pretty standard stuff, i.e. a playlist and video files.

**Comment by stephenbasile** [comment]
Would anyone be interested in helping maintain a fork of this? I considered it, since I've already contributed so much code to this repo, but I'm not sure I have the time by myself.

**Comment by jordanbaucke** [comment]
@stephenbasile I would. I'd probably be more help with a UI / web-management interface which is a nice-to-have but I've written a few modules to upload stills to S3 and other's nice to have. If we built-up a set of functionality to extend the actions and event handlers, as well as some interface to the logging I think people would find it interesting.

**Comment by stephenbasile** [comment]
@jordanbaucke Cool, making this competitive with commercial media servers would definitely require a web interface. As well as Playready support, Common Encryption (and other DRM), on the fly VOD packaging, etc. I'm kind of interested in doing all of that. I've done it before, but only for employers as closed source. 

Before taking on squashing the giant pile of bugs and feature requests on this project I'd definitely want to know another C programmer is interested in helping.

**Comment by jordanbaucke** [comment]
Cheers, sounds good to me- I'll keep my eyes open as well. There's probably
a decent business to fund development/some bounties running a hosted
version of an open-source platform (like Wowza, jwPlayer, etc.) if we can
get it stable / intuitive enough

**Comment by RuriRyan** [comment]
Hello guys,
I have some experience with C++ (was the language I learned programming in) but i'm now working with Perl. Some time back, when I tried to program C++ under Linux, I was like writing only C code. I would love to help you 'reviving' (?) this project/repo but I am unsure if I would be a good help, since I don't know how to write nginx modules, neither have I any experience with writing a rtmp/streaming server. I think I need some good amount of time to dive into these things.

**Comment by winlinvip** [comment]
Why not try https://github.com/simple-rtmp-server/srs ?

**Comment by misiek08** [comment]
Already working on it locally.
22 maj 2015 16:50 "winlin" <notifications@github.com> napisał(a):

**Comment by junaidnasir** [comment]
@mknwebsolutions rtmp has min 2 sec delay but hls have way more delay than that plus moving all those files to S3 (CDN's origin) as they are generated is a big headache. how do you overcome that ?

**Comment by mknwebsolutions** [comment]
@junaidnasir yes it definitely is!  We utilize enhanced networking servers, so moving files to S3 is almost unnoticeable.

**Comment by junaidnasir** [comment]
@mknwebsolutions can you share which tool are you using to move the files ? or have you mounted s3 using s3fs ?

**Comment by mknwebsolutions** [comment]
@junaidnasir s3 via s3fs, but again the key is having enhanced networking enabled, won't work as well without it.

**Comment by VackerSimon** [comment]
@junaidnasir "rtmp has min 2 sec delay"? A friend recently measured the delay from my servers to be around 1.1 seconds.

**Comment by junaidnasir** [comment]
@VackerSimon yes you are right, i forgot the player's default buffer size. @mknwebsolutions oh thank you, In my case s3fs was eating up all the CPU (recording HLS with 2 sec chunk so a lot of files to be moved), will test it with enhanced networking

**Comment by misiek08** [comment]
Using nginx-rtmp-module or srs with s3 is like taking two best, cheap encoders and they buy 1000 Wowza licenses to stream for 1000 viewers. S3 costs are so big, I don't wonna even ask you how much money you lost because using it.

**Comment by mknwebsolutions** [comment]
@misiek08 nginx-rtmp, srs, and woza will all use the same bandwidth and have the same bandwidth cost on AWS... S3 will have an additional cost for PUTS and GETS but it's $.005 per 1000/req PUT and $.004 per 10000/req GET.  Sure, I'm loosing 90 cents a month lol.

**Comment by misiek08** [comment]
I say about storage and bandwidth costs.....

**Comment by mknwebsolutions** [comment]
@misiek08 any of them, you set playlist to 30 seconds for HLS, then use AWS S3, CDN, or direct from server, they all use the same bandwidth pricing out.  With a playlist of 30 seconds your storage per channel is maybe 10-20MB of immediately expiring files. So no, there's not really a storage cost. I don't understand your comment.

**Comment by misiek08** [comment]
Yes, I see we have a misunderstanding. I tried to say, S3 is one of most expensive CDN I know and using it with such great, free software is bad idea (of course, for rich ppl is great, but I care about costs).

**Comment by mknwebsolutions** [comment]
@misiek08 gotcha, well I wouldn't say it's the most expensive CDN.  It's actually the cheapest of the best CDNs that have datacenters throughout the world.  There are very cheap CDNs, but you get what you pay for.

**Comment by misiek08** [comment]
I use to get 1 in USA and 1 in EU and have 40% of costs. That's my point of view.

**Comment by winlinvip** [comment]
When not care about the latence of stream, recommend to use WebP2P system: https://github.com/simple-rtmp-server/srs/wiki/v2_EN_WebP2P
User can direclty use the FlashP2P feature for the stream, when use SRS as media origin server and delivery HLS.

**Comment by winlinvip** [comment]
50%+ costs off, when use WebP2P.

**Comment by paulashford1975** [comment]
@jordanbaucke interested in any additions you have made. I want to be able to stream live broadcast to a final media file on Amazon s3 or another remote location. Also I am assuming that new apps/feeds players could be dynamically added by using a control interface, thus the ability to add /category/stream name?

**Comment by misiek08** [comment]
If you are interested in any dedicated solution you can contact me by mail. I have solutions made on multiple projects (srs, nginx-rtmp, own nodejs and Go programs, even Wowza - but it's expensive) and multiple, cheaper than Amazon cloud operators. I still have not enough time (because of university) to make a website with offer, but in some time I hope I will publish ready product.

I can help with nginx-rtmp. I can rate my knowledge in C for 5-7/10 and in nginx module development 5/10. Other languages/technologies: Go, PHP, node.js, Perl, Bash, Java (not enterprise), so I can help a little. Of course HTML/CSS included :)

**Comment by paulashford1975** [comment]
Super, if anyone wants to chat about modules and features please contact me, i can also be found on Skype at paulashford1975.

**Comment by misiek08** [comment]
@paulashford1975 can your solution hold 1kk+ chat users online? I made own chat too, but when going over 1kk users there are problems with realtime and delay shows up.

1kk users, about 500-1k messages/s constant, peak is 5-7k messages/s.

**Comment by btecu** [comment]
@sergey-dryabzhinsky seems very active and knowledgeable, maybe he'd be interested in this?
There are a lot of pull requests that should be merged, but the author is not responding to any of them.

**Comment by jordanbaucke** [comment]
Just as an aside: I've been working with this module for about a year and half. If anyone is interested in some contract C++ work in this area - desktop apps, rtmp and video stream ping me: jordan dot baucke at gmail dot com

**Comment by winlinvip** [comment]
Why not try https://github.com/simple-rtmp-server/srs , an alive project?

**Comment by btecu** [comment]
@winlinvip will try it once 2.0 is released.

**Comment by denji** [comment]
https://github.com/sergey-dryabzhinsky/nginx-rtmp-module

**Comment by winlinvip** [comment]
@btecu SRS is dead too.

**Comment by antonioparisi** [comment]
@mknwebsolutions and @jordanbaucke I like your solution of ingesting RTMP and produce in output HLS to S3 over CloudFront. My question though is how do you scale the nginx part in case the traffic gets higher? Also I was figuring out what is the best way to scale in case those nginx ec2 machines. I was thinking of docker maybe? Because you need to install nginx with rtmp module support and also configure S3FS to mount the S3 path where to output the HLS files.

**Comment by misiek08** [comment]
You can proxy connections by 3rd party tcp load balancer to many nginx instances. Scaling by docker is good solution.
Contact me if you are interested in some custom solutions, I have experience in live events (so far only up to 30.000 viewers, but there is no limit). I don't use AWS, because it's extremely expensive.

**Comment by antonioparisi** [comment]
@misiek08 so, streaming RTMP both upstream and downstream because low latency is required how would you scale it based on the traffic/connections using a load balancer or whatever above nginx?

**Comment by stephenbasile** [comment]
Not really tenable for outbound because you'd just be creating a single point of failure. For inbound, HAProxy is great for load balancing.

Spreading outbound RTMP traffic among a bunch of servers is really the central challenge of it. Unlike HTTP you can't move a client from one server to another, so they can get overloaded, etc. Usually it is done with DNS as far as I know. I work for a CDN that only does HLS/DASH/HDS streaming, so I'm a little behind on RTMP scaling.

**Comment by misiek08** [comment]
Haproxy on ingest side and edge servers for viewers. RTMP is bad protocol
on many levels and scaling it requires a lot of imagination. You have to
make a tree of servers. So one master (ingest) have 2 or more pulling
servers, then there are many pulling from pullers where users connects. You
shouldn't relay on DNS for load balancing. Just provide subdomain for each
server from last level and on website or in application provide different
address depending on connection count on those servers.

**Comment by jordanbaucke** [comment]
@all while I've got a long channel going about reducing latency - I'm interested in learning more about how beam.pro implements their "FTL" (Faster than Light) streaming- https://github.com/WatchBeam (they have a number of repos). It appears they're using a modified version of OBS + RTMP Ingest proxied (?) to a WebRTC Gateway? 

Anyone with any insight on this ?

**Comment by misiek08** [comment]
It's WebRTC, and it's not magic or their unique idea. There are many
examples on the Web for a long time now and it just works. You can engage
p2p too, to make latency even smaller and to handle scale without
extraordinary big bandwidth capacity on your (server) side.

**Comment by winlinvip** [comment]
It's time to introduce [SRS](https://github.com/ossrs/srs).

## Features

|   Feature     |   SRS     |   NGINX   | CRTMPD    | AMS       |   WOWZA   |    Janus  |
|   ----------- |   ------- |   -----   | --------- | --------  |   ------  |   ------  |
|   RTMP        |   Stable  |   Stable  |   Stable  |   Stable  |   Stable  |   X       |
|   WebRTC      |Experiment |   X       |   X       |   X       |   X       |   Stable  |
|   HLS         |   Stable  |   Stable  |   X       |   Stable  |   Stable  |   X       |
|   HTTP FLV    |   Stable  |   X       |   X       |   X       |   X       |   X       |
|   HLS(aonly)  |   Stable  |   X       |   X       |   Stable  |   Stable  |   X       |
|   HDS         | Experiment|   X       |   X       |   Stable  |   Stable  |   X       |
|   MPEG-DASH   | Experiment|   X       |   X       |   X       |   X       |   X       |
|   SRT         | Experiment|   X       |   X       |   X       |   Stable  |   X       |
|   HTTP Server |   Stable  |   Stable  |   X       |   X       |   Stable  |   X       |
|   GB28181     | Experiment|   X       |   X       |   X       |   X       |   X       |

## RTMP `SRS = 3 * NginxRTMP`

The data for playing RTMP was benchmarked by [srs-bench](https://github.com/ossrs/srs-bench):


|   Update      |    SRS    |    Clients    |     Type      |    CPU    |  Memory   | Commit        |
| ------------- | --------- | ------------- | ------------- | --------- | --------  | ------------  |
|   2014-12-07  |   2.0.67  |   10k(10000)  |   players     |   95%     |   656MB   |   [code][p12] |
|   2014-12-05  |   2.0.57  |   9.0k(9000)  |   players     |   90%     |   468MB   |   [code][p11] |
|   2014-12-05  |   2.0.55  |   8.0k(8000)  |   players     |   89%     |   360MB   |   [code][p10] |
|   2014-11-22  |   2.0.30  |   7.5k(7500)  |   players     |   87%     |   320MB   |   [code][p9]  |
|   2014-11-13  |   2.0.15  |   6.0k(6000)  |   players     |   82%     |   203MB   |   [code][p8]  |
|   2014-11-12  |   2.0.14  |   3.5k(3500)  |   players     |   95%     |   78MB    |   [code][p7]  |
|   2014-11-12  |   2.0.14  |   2.7k(2700)  |   players     |   69%     |   59MB    |   -           |
|   2014-11-11  |   2.0.12  |   2.7k(2700)  |   players     |   85%     |   66MB    |   -           |
|   2014-11-11  |   1.0.5   |   2.7k(2700)  |   players     |   85%     |   66MB    |   -           |
|   2014-07-12  |   0.9.156 |   2.7k(2700)  |   players     |   89%     |   61MB    |   [code][p6]  |
|   2014-07-12  |   0.9.156 |   1.8k(1800)  |   players     |   68%     |   38MB    |   -           |
|   2013-11-28  |   0.5.0   |   1.8k(1800)  |   players     |   90%     |   41M     |   -           |

## RTC `SRS = 2.5 * Janus`

The RTC benchmark data, by [srs-bench](https://github.com/ossrs/srs-bench/tree/feature/rtc#usage):

|   Update      |    Server     | Clients |    Type  |   CPU    | Memory   | Threads | Commit |
| ------------- | ------------  | ----- | ---------- | -------- | -------- | ----- | --------- |
|   2021-04-07  |   SRS/5.0.3   | 10000 | publishers | ~90% x 32 |  28GB  | 33   | [#2188](https://github.com/ossrs/srs/issues/2188) |
|   2021-04-07  |   SRS/5.0.3   | 3400  | publishers | ~95% x 8 |  6.3GB  | 12   | [#2188](https://github.com/ossrs/srs/issues/2188) |
|   2021-04-07  | Janus/0.10.10 | 1100  | publishers | ~55% x 8 |  169MB  | 1120  | [Crashed](https://github.com/meetecho/janus-gateway/issues/2626) |
|   2021-04-07  |   SRS/5.0.3   | 2000  | publishers | ~95% x 4 |  4.1GB  | 8    | [#2188](https://github.com/ossrs/srs/issues/2188) |
|   2021-03-31  |   SRS/5.0.2   | 1400  | publishers | ~90% x 4 |  3.1GB  | 6    | [#2188](https://github.com/ossrs/srs/issues/2188) |
|   2021-03-31  |   SRS/5.0.2   | 1400  | players    | ~93% x 4 |  1.0GB  | 6    | [#2188](https://github.com/ossrs/srs/issues/2188) |
|   2021-03-31  | Janus/0.10.10 | 700   | publishers | ~80% x 4 |  142MB  | 720  | |
|   2021-03-31  | Janus/0.10.10 | 700   | players    | ~82% x 4 |  283MB  | 720  | |
|   2021-03-31  |   SRS/4.0.87  | 550   | publishers | ~86% x 1 |  1.3GB  | 1    | |
|   2021-03-31  |   SRS/4.0.87  | 800   | players    | ~94% x 1 |  444MB  | 1    | |

> Note: The benchmark tool for Janus is [srs-bench](https://github.com/ossrs/srs-bench/tree/feature/rtc#janus), and startup script by [janus-docker](https://github.com/winlinvip/janus-docker#usage).
